token-erc-721/chaincode-javascript/lib/tokenERC721.js

```diff
@@ -0,0 +1,403 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+*/
+
+'use strict';
+
+const { Contract } = require('fabric-contract-api');
+
+// Define objectType names for prefix
+const balancePrefix = 'balance';
+const nftPrefix = 'nft';
+const approvalPrefix = 'approval';
+
+// Define key names for options
+const nameKey = 'name';
+const symbolKey = 'symbol';
+
+class TokenERC721Contract extends Contract {
+
+    /**
+     * BalanceOf counts all non-fungible tokens assigned to an owner
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} owner An owner for whom to query the balance
+     * @returns {Number} The number of non-fungible tokens owned by the owner, possibly zero
+     */
+    async BalanceOf(ctx, owner) {
+        // Execute a partial composite key query on all non-fungible tokens owned by the given owner
```

Comment:
Since this is early in the file, in the comment, explain the concept of the composite key and query, e.g.:
There is a key record for every nft in the format of
balancePrefix.owner.tokenId.
BalanceOf queries for and counts all records matching balancePrefix.owner.*